data/abi/abi_value.go

```diff
@@ -0,0 +1,304 @@
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package abi
+
+import (
+	"fmt"
+	"math/big"
+)
+
+// Value struct is the ABI Value, holding ABI Type information and the ABI value representation.
+type Value struct {
+	ABIType Type
+	value   interface{}
+}
+
+// MakeUint8 takes a go `uint8` and gives an ABI Value of ABI type `uint8`.
+func MakeUint8(value uint8) (Value, error) {
+	bigInt := big.NewInt(int64(value))
+	return MakeUint(bigInt, 8)
+}
+
+// MakeUint16 takes a go `uint16` and gives an ABI Value of ABI type `uint16`.
+func MakeUint16(value uint16) (Value, error) {
+	bigInt := big.NewInt(int64(value))
+	return MakeUint(bigInt, 16)
+}
+
+// MakeUint32 takes a go `uint32` and gives an ABI Value of ABI type `uint32`.
+func MakeUint32(value uint32) (Value, error) {
+	bigInt := big.NewInt(int64(value))
+	return MakeUint(bigInt, 32)
+}
+
+// MakeUint64 takes a go `uint64` and gives an ABI Value of ABI type `uint64`.
+func MakeUint64(value uint64) (Value, error) {
+	bigInt := big.NewInt(int64(0)).SetUint64(value)
+	return MakeUint(bigInt, 64)
+}
+
+// MakeUint takes a big integer representation and a type bitSize,
+// and returns an ABI Value of ABI Uint<bitSize> type.
+func MakeUint(value *big.Int, size uint16) (Value, error) {
+	typeUint, err := MakeUintType(size)
+	if err != nil {
+		return Value{}, err
+	}
+	upperLimit := big.NewInt(0).Lsh(big.NewInt(1), uint(size))
+	if value.Cmp(upperLimit) >= 0 {
+		return Value{}, fmt.Errorf("passed value larger than uint bitSize %d", size)
+	}
+	return Value{
+		ABIType: typeUint,
+		value:   value,
+	}, nil
+}
+
+// MakeUfixed takes a big rational number representation, a type bitSize, and a type precision,
+// and returns an ABI Value of ABI UFixed<bitSize>x<precision>
+func MakeUfixed(value *big.Int, size uint16, precision uint16) (Value, error) {
+	ufixedValueType, err := MakeUfixedType(size, precision)
+	if err != nil {
+		return Value{}, err
+	}
+	uintVal, err := MakeUint(value, size)
+	if err != nil {
+		return Value{}, err
+	}
+	uintVal.ABIType = ufixedValueType
+	return uintVal, nil
+}
+
+// MakeString takes a string and returns an ABI String type Value.
+func MakeString(value string) Value {
+	return Value{
+		ABIType: MakeStringType(),
+		value:   value,
+	}
+}
+
+// MakeByte takes a byte and returns an ABI Byte type value.
+func MakeByte(value byte) Value {
```

Comment:
Do we have a byte type and a uint8 type?

Reply:
byte type is just an alias of uint8. If it's for the sake of consistency, I can switch that to uint8.

Reply:
I suppose we want it mainly so we can have byte[32], instead of uint8[16].  I suppose that's fine. It's slightly odd that you can have two methods f(byte) and f(uint8), but I guess that's the same problem we have with an alias (we have a few, I guess).